File: base58.py
alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz"

max_int = 0x1ffffffffff
max_token = "zmM9z4E"


def encode(input: int) -> str:
    if input < 0 or input > max_int:
        raise ValueError(f"Expected integer between 0 and {max_int}, got {input}")

    result = ""
    num = input
    for i in range(7):
        num, rem = divmod(num, 58)
        result = alphabet[rem] + result
    assert num == 0
    return result


def decode(input: str) -> int:
    if len(input) != 7:
        raise ValueError(
            f"Expected token of length 7, got token of length {len(input)}"
        )

    if input > max_token:
        raise ValueError("Token is too big to decode into a 64-bit integer", input)

    result = 0
    for ch in input:
        result *= 58
        index = alphabet.index(ch)
        result += index
    return result

File: php/src/Base58.php
<?php

namespace Sjord\Fecid;

use \Exception;

final class Base58 {

    static $alphabet = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
    static $max_int = 0x1ffffffffff;
    static $max_token = "zmM9z4E";

    public static function encode($input) {
        global $alphabet, $max_int;
        if ($input < 0 || $input > static::$max_int) {
            throw new Exception("Expected integer between 0 and " . $max_int . ", got " . $input);
        }

        $result = "";
        $num = $input;
        for ($i = 0; $i < 7; $i++) {
            $rem = $num % 58;
            $num = intdiv($num, 58);
            $result = static::$alphabet[$rem] . $result;
        }
        assert($num == 0);
        return $result;
    }

    public static function decode($input) {
        if (strlen($input) != 7) {
            throw new Exception("Expected token of length 7, got token of length " . strlen($input));
        }

        if ($input > static::$max_token) {
            throw new Exception("Token is too big to decode into a 64-bit integer");
        }

        $result = 0;
        for ($i = 0; $i < strlen($input); $i++) {
            $result *= 58;
            $index = strpos(static::$alphabet, $input[$i]);
            $result += $index;
        }
        return $result;
    }
}
